transformer_engine/pytorch/csrc/extensions/attention.cu

```diff
@@ -510,6 +511,102 @@ std::vector<py::object> fused_attn_bwd(
   return {py_dQ, py_dK, py_dV, py::cast(dBias)};
 }
 
+/***************************************************************************************************
+ * Support BSHD, SBHD, and THD formats for Context Parallel: Fused out correction in forward
+ **************************************************************************************************/
+
+template <typename dtype, bool causal>
+void fused_out_correction_helper(at::Tensor &out, const std::vector<at::Tensor> &out_per_step,
+                                 const at::Tensor &lse, const std::vector<at::Tensor> &lse_per_step,
+                                 const py::object &cu_seqlens, std::string qkv_format, int cp_size,
+                                 int rank, bool softmax_lse_in_packed_format) {
+  std::unordered_map<std::string, NVTE_QKV_Format> format_map = {
+      {"sbhd", NVTE_QKV_Format::NVTE_SBHD},
+      {"bshd", NVTE_QKV_Format::NVTE_BSHD},
+      {"thd", NVTE_QKV_Format::NVTE_THD}};
+
+  int lse_seqlen;
+  int batch;
+  int num_heads;
+  int dim_per_head;
+  int total_tokens;
+  int cu_seqlens_size = 0;
+  int *cu_seqlens_ptr = nullptr;
+  if (qkv_format == "sbhd") {
+    lse_seqlen = out.size(0);
+    batch = out.size(1);
+    num_heads = out.size(2);
+    dim_per_head = out.size(3);
+    total_tokens = lse_seqlen * batch;
+  } else if (qkv_format == "bshd") {
+    lse_seqlen = out.size(1);
+    batch = out.size(0);
+    num_heads = out.size(2);
+    dim_per_head = out.size(3);
+    total_tokens = lse_seqlen * batch;
+  } else if (qkv_format == "thd") {
+    total_tokens = out.size(0);
+    num_heads = out.size(1);
+    dim_per_head = out.size(2);
+    batch = cu_seqlens.cast<at::Tensor>().size(0) - 1;
+    cu_seqlens_size = sizeof(int) * (batch + 1);
+    cu_seqlens_ptr = cu_seqlens.cast<at::Tensor>().data_ptr<int>();
+    if (softmax_lse_in_packed_format) {
+      lse_seqlen = total_tokens;
```

Comment:
this needs special consideration, because TE does token count quantization which could make `lse_seqlen >= total_tokens`, also make `half_lse_seqlen >= lse_seqlen / 2`. (refer this PR #1396 ).